### PR TITLE
[Feature] Graceful handling of exceptions thrown within OpenMP blocks

### LINKED ISSUE #3338: [Random Walk] `dgl.sampling.random_walk()` has several unexpected behaviors/segfaults when given seed node ids outside of the graph node id range
## 🐛 Bug

When I call `dgl.sampling.random_walk(graph, seed_node_ids)` with seed node ids that are not valid (eg outside of the range of node ids owned by the input graph), I am seeing the following unexpected behaviors:
- segmentation faults, 
- Abort with a C++ stack traces
- (worryingly) sometimes the function succeeds and returns a trace of all -1's. 

Interestingly, the actual value of the invalid seed_node_id can reliably trigger each of the above conditions. This sounds like an OOB C++ bug.
Perhaps the C++ random walk code should do some sanity checks on the input seed node ids, and early-terminate when it detects invalid inputs (ideally with a Python Exception, rather than a C++ Exception + Abort?).

## To Reproduce

Here is a repro script: https://gist.github.com/erickim555/77045b3b22dae797c9393a5161a3cd71#file-dgl_randwalk_node_id_out_of_bounds_error-py

## Expected behavior

The DGL code should reliably detect invalid input seed node ids and error early, rather than segfault/undefined-behavior/etc.

## Environment

 - DGL Version (e.g., 1.0): a fairly recent nightly version, `'0.8a210818'`
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3):
 - OS (e.g., Linux): Ubuntu 16.04
 - How you installed DGL (`conda`, `pip`, source): nightly
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version (if applicable):
 - GPU models and configuration (e.g. V100):
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
This PR ports PyTorch's OpenMP C++ exception handling so that DGL will not crash if a managed C++ exception (e.g. `LOG(FATAL)`, `CHECK`, etc.) is thrown.

Also added boundary checks in OpenMP random walks as a demonstration.  This also fixes #3338 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by BarclayII** [comment]
@dgl-bot run all c5.9xlarge dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run all c5.9xlarge dmlc/master

**Comment by BarclayII** [comment]
@dgl-bot run c5.9xlarge all dmlc/master

**Comment by BarclayII** [comment]
Seems that the impact is minimal.  Although `from_scipy` with small COO matrix slowed down by 100% I think that's OK because the base is already 8e-5s.

**Comment by VoVAllen** [comment]
LGTM

File: src/graph/sampling/randomwalks/node2vec_randomwalk.h [hunk 1]
```diff
@@ -163,7 +163,7 @@ std::pair<IdArray, IdArray> Node2vecRandomWalk(
                                                   edges, csr_has_data, prob, terminate);
     };
 
-  return GenericRandomWalk<XPU, IdxType>(seeds, max_num_steps, step);
+  return GenericRandomWalk<XPU, IdxType>(seeds, max_num_steps, step, g->NumVertices(0));
```
Comment (VoVAllen) [inline src/graph/sampling/randomwalks/node2vec_randomwalk.h:166]: Is this always etype 0?

Comment (BarclayII) [reply src/graph/sampling/randomwalks/node2vec_randomwalk.h:166]: Yes.  Node2vec currently only supports homogeneous graphs.

**Pull request merged**